addons/account/models/account_reconcile_model.py

```diff
@@ -355,267 +367,372 @@ def _prepare_reconciliation(self, st_line, aml_ids=[], partner=None):
-
-                -- Determine a matching or not with the statement line communication using the aml.name, move.name or move.ref.
+        # N.B: 'communication_flag' is there to distinguish invoice matching through the number/reference
+        # (higher priority) from invoice matching using the partner (lower priority).
+        query = r'''
+        SELECT
+            %(sequence)s                        AS sequence,
+            %(model_id)s                        AS model_id,
+            st_line.id                          AS id,
+            aml.id                              AS aml_id,
+            aml.currency_id                     AS aml_currency_id,
+            aml.date_maturity                   AS aml_date_maturity,
+            aml.amount_residual                 AS aml_amount_residual,
+            aml.amount_residual_currency        AS aml_amount_residual_currency,
+            aml.balance                         AS aml_balance,
+            aml.amount_currency                 AS aml_amount_currency,
+            account.internal_type               AS account_internal_type,
+
+            -- Determine a matching or not with the statement line communication using the aml.name, move.name or move.ref.
+            (
+                aml.name IS NOT NULL
+                AND
+                substring(REGEXP_REPLACE(aml.name, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*') != ''
+                AND
+                    regexp_split_to_array(substring(REGEXP_REPLACE(aml.name, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*'),'\s+')
+                    && regexp_split_to_array(substring(REGEXP_REPLACE(st_line.payment_ref, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*'), '\s+')
+            )
+            OR
+                regexp_split_to_array(substring(REGEXP_REPLACE(move.name, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*'),'\s+')
+                && regexp_split_to_array(substring(REGEXP_REPLACE(st_line.payment_ref, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*'), '\s+')
+            OR
+            (
+                move.ref IS NOT NULL
+                AND
+                substring(REGEXP_REPLACE(move.ref, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*') != ''
+                AND
+                    regexp_split_to_array(substring(REGEXP_REPLACE(move.ref, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*'),'\s+')
+                    && regexp_split_to_array(substring(REGEXP_REPLACE(st_line.payment_ref, '[^0-9|^\s]', '', 'g'), '\S(?:.*\S)*'), '\s+')
+            )                                   AS communication_flag,
+            -- Determine a matching or not with the statement line communication using the move.payment_reference.
+            (
+                move.payment_reference IS NOT NULL
+                AND
+                regexp_replace(move.payment_reference, '\s+', '', 'g') = regexp_replace(st_line.payment_ref, '\s+', '', 'g')
+            )                                   AS payment_reference_flag
```

Comment:
maybe we could split this query in two parts: matching on communication_flag & matching on payment_reference_flag to avoid redundant code in `select` and `where`

Reply:
You mean, making two queries ? I'm afraid perf wouldn't like that :/

Reply:
@smetl @oco-odoo 
Can you please guys consider this PR https://github.com/odoo/odoo/pull/50413 and extract the select in a separated function? :pray: 
We needed to be able to customize the behaviour implied by this two flags to avoid the issue described in https://github.com/odoo/odoo/issues/43291 
Although this particular issue should be fixed in the next version, would be great if we can keep Odoo core standard and don't have to rely on this kind of PR for another issue... :no_mouth:

Reply:
@grindtildeath I just r+ed your PR :) It'll be applied to my refactoring as it gets forward-ported (or when this branch is rebased on master, depending when we merge it ^^). This way you'll still be able to customize the flags if you need to :)